File: leetcode/373-find-k-pairs-with-smallest-sums/main.go
package main

import (
	"container/heap"
	"fmt"
	"sort"
)

/*
	1st approach
	- use a priority queue

	Time 	O(nlogn)
	Space	O(n)
	188 ms, faster than 18.75%
*/
func kSmallestPairs(nums1 []int, nums2 []int, k int) [][]int {
	pq := &PriorityQueue{}
	heap.Init(pq)
	cnt := 0
	for _, a := range nums1 {
		for _, b := range nums2 {
			heap.Push(pq, &Item{a + b, a, b, cnt})
			cnt++
		}
	}
	i := 0
	min := 0
	if pq.Len() < k {
		min = pq.Len()
	} else {
		min = k
	}
	res := [][]int{}
	for i < min {
		pop := heap.Pop(pq).(*Item)
		res = append(res, []int{pop.Nums1, pop.Nums2})
		i++
	}
	return res
}

/*
	Implement Priority Queue
*/

type Item struct {
	Product int // The priority of the item in the queue.
	Nums1   int
	Nums2   int
	// The index is needed by update and is maintained by the heap.Interface methods.
	Index int // The index of the item in the heap.
}

// A PriorityQueue implements heap.Interface and holds Items.
type PriorityQueue []*Item

func (pq PriorityQueue) Len() int { return len(pq) }

func (pq PriorityQueue) Less(i, j int) bool {
	// We want Pop to give us the highest, not lowest, priority so we use greater than here.
	return pq[i].Product < pq[j].Product
}

func (pq PriorityQueue) Swap(i, j int) {
	pq[i], pq[j] = pq[j], pq[i]
	pq[i].Index = i
	pq[j].Index = j
}

func (pq *PriorityQueue) Push(x interface{}) {
	n := len(*pq)
	item := x.(*Item)
	item.Index = n
	*pq = append(*pq, item)
}

func (pq *PriorityQueue) Pop() interface{} {
	old := *pq
	n := len(old)
	item := old[n-1]
	item.Index = -1 // for safety
	*pq = old[0 : n-1]
	return item
}

/*
	2nd approach
	- use built-in sort

	Time 	O(nlogn)
	Space	O(n)
	192 ms, faster than 12.50%
*/
func kSmallestPairs1(nums1 []int, nums2 []int, k int) [][]int {
	arr := []*Item{}
	for _, a := range nums1 {
		for _, b := range nums2 {
			o := &Item{a + b, a, b, 0}
			arr = append(arr, o)
		}
	}
	sort.Slice(arr, func(i, j int) bool {
		return arr[i].Product < arr[j].Product
	})
	min := 0
	if len(arr) < k {
		min = len(arr)
	} else {
		min = k
	}
	res := [][]int{}
	i := 0
	for i < min {
		res = append(res, []int{arr[i].Nums1, arr[i].Nums2})
		i++
	}
	return res
}

func main() {
	nums1 := []int{1, 7, 11}
	nums2 := []int{2, 4, 6}
	fmt.Println(kSmallestPairs(nums1, nums2, 3))
	fmt.Println(kSmallestPairs1(nums1, nums2, 3))
}

File: leetcode/373-find-k-pairs-with-smallest-sums/main.py
"""
    1st: brute force sort

	Time 	O(AB logAB)
	Space	O(MN)
	260 ms, faster than 34.73%
"""


class Solution:
    def kSmallestPairs(self, nums1: List[int], nums2: List[int], k: int) -> List[List[int]]:
        arr = []
        for x in nums1:
            for y in nums2:
                arr.append((x + y, x, y))
        arr.sort()
        arr = arr[:k]
        res = []
        for i in range(len(arr)):
            s, x, y = arr[i]
            res.append([x, y])
        return res


"""
    2nd: min heap, learned from others
    - same logic as 786
    - similar to lc23, 373, 378, 786
    - since the arrays are sorted, we dont have to consider all pairs
    we can just care if we can pop k smallest items out from the stack
    
    e.g.
    a = [1, 3, 11]
    b = [2, 4, 6]
    k = 4

    heap = [1+2, 3+2, 11+2], consider the total of every element of nums1 + nums2[0]
    
    heap = [3, 5, 13]
    
    count = 0, we pop the first one because it must be smallest one,
    3 (i=0, j=0) = heap.pop()
    heap = [5, 11]
    then we consider adding nums1[0] + nums2[0+1 = 1] = 1 + 4 = 5 to the heap
    heap = [5, 5, 13]
    
    count = 1, we pop the first one because it must be smallest one,
    5 (i=0, j=1) = heap.pop()
    heap = [5, 13]
    then we consider adding nums1[0] + nums2[1+1] = 1 + 6 = 7 to the heap
    heap = [5, 7, 13]

    count = 2, we pop the first one because it must be smallest one,
    5 (i=1, j=0) = heap.pop()
    heap = [7, 13]
    then we consider adding nums1[1] + nums2[0+1] = 3 + 4 = 7 to the heap
    heap = [7, 7, 13]

    count = 3 we pop the first one because it must be smallest one,
    7 (i=0, j=2) = heap.pop()
    since count + 1 == k, we are done!!! 7 is the 4th smallest pairsum

    ref:
    - https://leetcode.com/problems/find-k-pairs-with-smallest-sums/discuss/84550/Slow-1-liner-to-Fast-solutions
    - https://www.youtube.com/watch?v=APZbA_q1zAc

	Time 	O(A + klogA)
	Space	O(MN)
	36 ms, faster than 86.84%
"""


class Solution(object):
    def kSmallestPairs(self, nums1, nums2, k):
        if len(nums1) == 0 or len(nums2) == 0 or k == 0:
            return []
        pq = []
        for i in range(len(nums1)):
            x = nums1[i]
            y = nums2[0]
            heapq.heappush(pq, (x+y, i, 0))
        res = []
        while len(pq) > 0:
            _, i, j = heapq.heappop(pq)
            res.append([nums1[i], nums2[j]])
            if len(res) == k:
                break
            if j + 1 < len(nums2):
                x = nums1[i]
                y = nums2[j+1]
                heapq.heappush(pq, (x+y, i, j+1))
        return res
